pallets/gear/src/lib.rs

```diff
@@ -260,14 +265,60 @@ pub mod pallet {
             .ok()
         }
 
+        /// Returns true if a program has been successfully initialized
+        pub fn is_initialized(program_id: H256) -> bool {
+            common::get_program_state(program_id)
+                .map(|s| matches!(s, ProgramState::Initialized))
+                .unwrap_or(false)
+        }
+
         /// Returns true if a program resulted in an error during initialization
         /// but hasn't been explicitly removed from storage by its creator
-        pub fn is_uninitialized(program_id: H256) -> bool {
+        pub fn is_failed(program_id: H256) -> bool {
             ProgramsLimbo::<T>::get(program_id)
                 .map(|_| true)
                 .unwrap_or(false)
         }
 
+        /// Internal helper function. Moves messages from waiting list to the queue. Returns a flag if there are
+        /// more (probably) messages and remained weight.
+        fn requeue_waiting_messages(
+            program_id: H256,
+            remaining_weight: Weight,
+        ) -> (Option<WaitingMessages>, Weight) {
+            let one_step_weight = T::DbWeight::get().reads(2);
+            let message_count = remaining_weight / one_step_weight;
+
+            let actual_count = common::WaitingMessageIterator::drain(program_id, None)
+                .take(message_count as usize)
+                .fold(0u64, |acc, (m, _)| {
+                    common::queue_message(m);
+                    acc + 1
+                });
+
+            let result_weight = remaining_weight - actual_count * one_step_weight;
+            if actual_count < message_count {
```

Comment:
How about the case when `actual_count` == `message_count` and there are no items left in the `WaitingMessageIterator`? That is there were exactly as many messages in the wait list as we could process due to weight constraint

Reply:
Edge-case =).

For the sake of simplicity I decided not to make additional storage read to find out if there are more message but just schedule requeueing in the next block.

Thanks for comment anyway - I wrongly calculate `one_step_wait`. There should also one storage read by `common::queue...`